Add build time initialization priority checking

fabiobaltieri opened the pull request:
Hi, put this together trying to fight the neverending battle against initialization priorities. 

This is how it works: after the build, scans through all the object files in the build directory looking for relocation sections matching the init format, the ones that looks like `.rel.z_init_POST_KERNEL90_`. These contains references to the device structure with the correct ordinal as that's passed as an argument to the init call, for example:

```
Relocation section '.rel.z_init_POST_KERNEL90_' at offset 0x4c94 contains 2 entries:
 Offset     Info    Type                Sym. Value  Symbol's Name
00000000  00001002 R_ARM_ABS32            00000001   led_gpio_init
00000004  00003202 R_ARM_ABS32            00000000   __device_dts_ord_19
```

And that's used to build a map of device ordinals and their actual init priority in the build:
```
(Pdb) pp self._dev_priorities
{11: <Priority PRE_KERNEL_1 40>,
 19: <Priority POST_KERNEL 90>,
 59: <Priority PRE_KERNEL_1 30>,
 96: <Priority PRE_KERNEL_1 55>}
```

Then the list is matched with EDT Nodes and compared against their `depends_on` lists (this is built up by edtlib based on the tree hierarchy and phandle properties), for example:

```
(Pdb) pp dev
<Node /leds/led_1 in 'build/zephyr/zephyr.dts.pre', binding dts/bindings/led/gpio-leds.yaml>
(Pdb) pp dev.depends_on
[<Node /leds in 'build/zephyr/zephyr.dts.pre', binding dts/bindings/led/gpio-leds.yaml>,
 <Node /soc/gpio@50000000 in 'build/zephyr/zephyr.dts.pre', binding dts/bindings/gpio/nordic,nrf-gpio.yaml>]
```

The resulting device pairs are then checked to ensure that they are initialized in the right sequence, the scripts emit INFO for every check in verbose mode, WARNING for dependent devices at the same priority (which can optionally return with an error code) and ERROR for incorrect priorities.

Examples:
```
$ west build -p -b nrf52dk_nrf52832 samples/hello_world/ -DCONFIG_LED=y
$ ./scripts/build/check_init_priorities.py -v
INFO: /leds POST_KERNEL 90 > /soc/gpio@50000000 PRE_KERNEL_1 40
INFO: /leds POST_KERNEL 90 > /soc/gpio@50000000 PRE_KERNEL_1 40
INFO: /leds POST_KERNEL 90 > /soc/gpio@50000000 PRE_KERNEL_1 40
INFO: /leds POST_KERNEL 90 > /soc/gpio@50000000 PRE_KERNEL_1 40
```

This one is reproducing: https://github.com/zephyrproject-rtos/zephyr/issues/55258#issuecomment-1454755400

```
$ git revert -n e72d1bf970892d99628532acf908c4966d3a02fe
$ west build -p -b mimxrt1060_evk samples/drivers/flash_shell
$ ./scripts/build/check_init_priorities.py -v
INFO: /soc/uart@40184000 PRE_KERNEL_1 55 > /soc/ccm@400fc000 PRE_KERNEL_1 30
WARNING: /soc/spi@402a8000/is25wp064@0 POST_KERNEL 50 == /soc/spi@402a8000 POST_KERNEL 50
```

Not too sure what's up with this one:

```
$ west build -p -b frdm_k22f samples/hello_world
$ ./scripts/build/check_init_priorities.py -v
INFO: /soc/uart@4006b000 PRE_KERNEL_1 55 > /soc/sim@40047000 PRE_KERNEL_1 30
ERROR: /soc/pinmux@4004d000 PRE_KERNEL_1 0 < /soc/sim@40047000 PRE_KERNEL_1 30
ERROR: /soc/pinmux@4004c000 PRE_KERNEL_1 0 < /soc/sim@40047000 PRE_KERNEL_1 30
ERROR: /soc/pinmux@4004b000 PRE_KERNEL_1 0 < /soc/sim@40047000 PRE_KERNEL_1 30
ERROR: /soc/pinmux@4004a000 PRE_KERNEL_1 0 < /soc/sim@40047000 PRE_KERNEL_1 30
ERROR: /soc/pinmux@40049000 PRE_KERNEL_1 0 < /soc/sim@40047000 PRE_KERNEL_1 30
INFO: /soc/gpio@400ff100 POST_KERNEL 40 > /soc/pinmux@4004d000 PRE_KERNEL_1 0
INFO: /soc/gpio@400ff0c0 POST_KERNEL 40 > /soc/pinmux@4004c000 PRE_KERNEL_1 0
INFO: /soc/gpio@400ff080 POST_KERNEL 40 > /soc/pinmux@4004b000 PRE_KERNEL_1 0
INFO: /soc/gpio@400ff040 POST_KERNEL 40 > /soc/pinmux@4004a000 PRE_KERNEL_1 0
INFO: /soc/gpio@400ff000 POST_KERNEL 40 > /soc/pinmux@40049000 PRE_KERNEL_1 0
```

Another interesting case, this is problematic cause it may make sense to initialize `phy` and `mdio` before the mac:

```
$ west build -p -b sam_v71_xult samples/net/sockets/echo_client
INFO: check_init_priorities build_dir: /usr/local/google/home/fabiobaltieri/zephyrproject/zephyr/build/zephyr/..
ERROR: /soc/ethernet@40050000/phy POST_KERNEL 70 < /soc/ethernet@40050000 POST_KERNEL 80
INFO: /soc/ethernet@40050000/phy POST_KERNEL 70 > /soc/ethernet@40050000/mdio POST_KERNEL 60
ERROR: /soc/ethernet@40050000/mdio POST_KERNEL 60 < /soc/ethernet@40050000 POST_KERNEL 80
```

Looked at the code and I think it does not really matter in this case, but I added some code to swap the check for specific pairs of compatibles just in case it does.

The check can be integrated with the build (`CONFIG_CHECK_INIT_PRIORITIES`) with the option of failing or not on warnings.

Main limitation: this relies on the devicetree to find actual dependencies, so any dependency implemented using things like `device_get_binding`, nodelabels, compatible lookup, path lookup etc is not going to show up. I tried pretty hard to infer dependencies using other  object relocation sections and symbols but it was getting way too many false positives to be useful.

In this form I think this should have few enough false positives that it may be feasible to fix those or exclude the check from some tests (like the build_all ones) and have this run in CI to catch potential regression (ahemm... https://github.com/zephyrproject-rtos/zephyr/pull/55048).

~What do you folks think? For the moment I'll leave this as RFC and keep a DNM patch that enables this by default in verbose mode while I go through the CI fails a bit at a time and fix more corner cases.~

I've done few passes of the CI run (see my comment below for the details), having this working in CI needs more work with few platform owners, but this seem to be good enough to be used to check application builds in most cases, so I'm proposing to review and merge it keeping it off by default.

Update:
- discussed at the Architecture WG
- added tests

fabiobaltieri committed:
scripts: build: add a script for build time init priority validation

Add a check_init_priorities.py scripts. This goes through all the object
files looking for known initialization sections and building a map of
device ordinals and their effective initialization priority. Then
compares that with the list of dependencies generated based on the
device tree and reports any warning (dependent devices, same priority)
or error (incorrect priority).

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
Kconfig.zephyr: add an option for checking build time init priority

Add an option for running the build time initialization priority check
as part of the build.

By default the build is going to fail if any error is reported
(dependent devices, inverted priority), optionally it can also be failed
on warnings (dependent devices, same priority).

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Hey, did few iteration checking this against the CI run, this now handles multi-build configs (sysbuild) and there's a list of known inverted priorities. On the last run with the check enabled by default it failed 88 targets in CI that have somethig with opposite priority than the devicetree dependency, this is what I found:

- bcm958402m2_m7: pcie and paxdma
- (frdm_k82f, twr_kv58f220m, frdm_k64f...) and most nxp ones: pinmux and sim (clock)
- mec172xmodular_assy6930: uart and espi (uart declared incorrectly in the dts?)
- mimxrt595_evk_cm33 samples/drivers/display: mipi and lcdif
- native_posix tests/subsys/emul: emul devices hardcoded priority, should probably just turn it off for the test
- samples/subsys/ipc/ipc_service/icmsg: ipc0 before mbox
- nrf5340dk_nrf5340_cpunet: ipc0,ipc1 before mbox

These would have to be fixed if there's interest in making the check run in CI, but for the moment it should be good enough to be useful downstream for actual applications, so I propose to have this merged as an optional check for now.

nordicjm commented:
> nrf5340dk_nrf5340_cpunet: ipc0,ipc1 before mbox

Ping @emob-nordic

fabiobaltieri commented:
> > nrf5340dk_nrf5340_cpunet: ipc0,ipc1 before mbox
> 
> Ping @emob-nordic

Sorry realized I bodged the list on those a bit, the exact tests that triggered are these: https://gist.github.com/fabiobaltieri/a147b80cbe59cb390b87110476d7a8c3, the error with node paths is

```
INFO: check_init_priorities build_dir: /usr/local/google/home/fabiobaltieri/zephyrproject/zephyr/build/zephyr/..
ERROR: /ipc/ipc1 POST_KERNEL 46 < /soc/mbox@41012000 POST_KERNEL 50
ERROR: /ipc/ipc0 POST_KERNEL 46 < /soc/mbox@41012000 POST_KERNEL 50
```

fabiobaltieri commented:
More context since I dropped the DNM patch to enable in CI: this is the full list of tests that was failing in the last run: https://gist.github.com/fabiobaltieri/99c0d032b74f1e756c408af5c62723a7. That is the testplan that was generated with this diff so there's probably more in practice.

nordicjm commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,306 @@
+        self.log = log
+
+        edtser = os.path.join(build_dir, "zephyr", "edt.pickle")
+        with open(edtser, "rb") as f:
```
Should this handle errors if the file doesn't exist?

nordicjm commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,306 @@
+        allow_abbrev=False)
+
+    parser.add_argument("-d", "--build-dir", default="build",
+                        help="build directory to create or use")
```
This can create a build directory?

fabiobaltieri replied:
Woops, copy paste...

nordicjm commented on CMakeLists.txt:
```diff
@@ -1715,6 +1715,18 @@ if(CONFIG_BUILD_OUTPUT_INFO_HEADER)
+    post_build_commands
+    COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/build/check_init_priorities.py
+        --build-dir ${PROJECT_BINARY_DIR}/..
+	${fail_on_warning}
```
alignment

fabiobaltieri commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,306 @@
+_SETTINGS_FILE = "zephyr_settings.txt"
+
+# List of compatibles for node where the initialization priority should be the
+# opposite of the device tree infered depdency.
```
Done.

bjarki-trackunit commented:
I hinted a bit at a neat feature I would love to see implemented in something like west, to at least visualize the init sequence, which can be used to validate that a driver/subsys is actually included in the build, and to manually validate the order. By altering the check_init_priorities.py script a bit, I can make it produce the following output

```
Boot sequence:
  0: atmel_samc_init:
    level: PRE_KERNEL_1 0

  1: init_mbox_module:
    level: PRE_KERNEL_1 30

  2: statics_init:
    level: PRE_KERNEL_1 30

  3: init_mem_slab_module:
    level: PRE_KERNEL_1 30

  4: gpio_sam0_init:
    level: PRE_KERNEL_1 40
    devices: __device_dts_ord_13, __device_dts_ord_8, __device_dts_ord_48

  5: sam0_eic_init:
    level: PRE_KERNEL_1 40
    devices: __device_dts_ord_26

  6: uart_sam0_init:
    level: PRE_KERNEL_1 55
    devices: __device_dts_ord_37, __device_dts_ord_34, __device_dts_ord_30

  7: uart_console_init:
    level: PRE_KERNEL_1 60

  8: sys_clock_driver_init:
    level: PRE_KERNEL_2 0

  9: k_sys_work_q_init:
    level: POST_KERNEL 40
```

It can't validate the subsystems, but it could be a useful intermediary tool to help with debugging :)

fabiobaltieri commented:
> I hinted a bit at a neat feature I would love to see implemented in something like west, to at least visualize the init sequence, which can be used to validate that a driver/subsys is actually included in the build, and to manually validate the order. By altering the check_init_priorities.py script a bit, I can make it produce the following output
> ...
> It can't validate the subsystems, but it could be a useful intermediary tool to help with debugging :)

Hey that's a cool idea! Should it be grouped by level? Like

```
level: PRE_KERNEL_1 0
  0: atmel_samc_init:

level: PRE_KERNEL_1 30
  1: init_mbox_module:
  2: statics_init:
  3: init_mem_slab_module:

level: PRE_KERNEL_1 40
  4: gpio_sam0_init:
    devices: __device_dts_ord_13, __device_dts_ord_8, __device_dts_ord_48
  5: sam0_eic_init:
    devices: __device_dts_ord_26
...
```

Regardless, I think it's a great addition and can see how one may want to have this generated in CI and compared against a golden copy of the validated sequence to catch unintentional changes.

Maybe let's sort this current feature set and have that as a follow up?

bjarki-trackunit commented:
> Maybe let's have the current feature set reviewed and do that as a follow up?

That would be the plan :) The foundational mechanism that sorts the init entries in this PR can be reused to do stuff like this in the future :)

fabiobaltieri committed:
scripts: build: add a test suite for check_init_priorities

Add a bunch of unit tests for the various check_init_priorities classes.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
ci: add a workflow for script unit tests

Add a "Scripts tests" workflow, for now just run pytest against
"scripts/build", but keep the naming generic so this could be reused for
other tests under scripts that are not covered by the other workflows.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Updates:
- added a testsuite for the script
- added a new CI workflow, went for "Scripts tests" with the idea that it could be reused for other tests in scripts (other build stuff, compliance...)
- dropped the RFC tag, feedback from the WG discussion were positive, don't be shy to review. :-)

fabiobaltieri committed:
doc: build: add a reference to check_init_priorities

Add a reference to the check_init_priorities script in the cmake
"Supporting Scripts and Tools" section.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

bjarki-trackunit commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,319 @@
+
+# List of compatibles for node where the initialization priority should be the
+# opposite of the device tree inferred dependency.
+_INVERTED_PRIORITY_COMPATIBLES = set([
```
I believe we should probably incentivize changing the devicetree nodes to comply with the init order in the future:

Changing around the devicetree a bit from this:
```
gmac: gmac {
	mdio: mdio {
	};
	phy: phy {
	};
};
```
to:
```
gmac: gmac {
    phy = <&phy>;
};
phy: phy {
    mdio = <&mdio>;
};
mdio: mdio {
};
```
should comply with this script. This way, we don't have to maintain this list of exceptions in the future.

fabiobaltieri replied:
That's certainly an option, certainly something that could be cool do in the same way on the various network driver. Let me remove that override so that the specific platform fails again, will still leave the code to invert the check in there since I suspect we'll use it for some case anyway. Then once the tool is merged we can open the conversation with the driver owners.

fabiobaltieri replied:
Dropped

tejlmand commented on scripts/build/check_init_priorities_test.py:
```diff
@@ -0,0 +1,336 @@
+from elftools.elf.relocation import RelocationSection
+from elftools.elf.sections import SymbolTableSection
+
+ZEPHYR_BASE = os.getenv("ZEPHYR_BASE")
+sys.path.insert(0, os.path.join(ZEPHYR_BASE, "scripts/build"))
```
it is not required to have ZEPHYR_BASE in environment, so please use `Path(__file__).resolve()` or `os.path.dirname(__file__)`.

fabiobaltieri replied:
...is this even needed?! Seems to work without it, let me try and drop it entirely.

fabiobaltieri replied:
Works, go figure... gone.

tejlmand reviewed (changes requested):
Thanks for this.

Overall impression is good :+1: 

A couple of minor details noticed.

fabiobaltieri commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,319 @@
+                      "APPLICATION", "SMP"]
+
+# Build project settings file, used to detect and skip sub-builds.
+_SETTINGS_FILE = "zephyr_settings.txt"
```
Yeah, not ideal, so I need something to identify a subbuild, initially used `build.ninja` but then it would not work with Makefile since that generates a Makefile for every subdirectory, was left with `zephyr_modules.txt` and `zephyr_settings.txt`. I could look for `CMakeCache.txt`? Or maybe add an explicit marker in the build directory, say adding a `touch ${PROJECT_BINARY_BUILD}/.zephyr_build` somewhere in the build system and look for that instead.

If we get to the point of enabling this in CI it would fail and catch it if module system gets reworked, but it'd be cool to not depend on that in the first place. Can you think of any better strategy?

tejlmand replied:
just to confirm my understanding.

You want to avoid processing sub-builds build directories, such as a TF-M build and similar, for performance reasons ?

Such sub-builds can be
1.  additional Zephyr builds
2. other CMake based builds, like [TF-M](https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/trusted-firmware-m/)
3. a Makefile based project, like [external_lib](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/external_lib).
4. Other / custom type projects

If this is the case, the `CMakeCache.txt` approach is safest cause that will catch 1 & 2, whereas current approach will only catch 1, and bear the built-in risk mentioned earlier.
3. might be present in a build but often represents integration to legacy systems, but we can't do much for those.
4. I wouldn't worry too much.

fabiobaltieri replied:
> just to confirm my understanding.
> 
> You want to avoid processing sub-builds build directories, such as a TF-M build and similar, for performance reasons ?

Yes but not for performance reasons, the script goes through all the object files looking for defined device structs and mapping their priority, if there are nested (Zephyr) projects they can potentially redefine the same device ordinal at different priority (or confuse the script in other ways). The script detects the double define (the two "device already defined" exceptions) but then does not know what to do with it.

> If this is the case, the `CMakeCache.txt` approach is safest cause that will catch 1 & 2, whereas current approach will only catch 1, and bear the built-in risk mentioned earlier. 3. might be present in a build but often represents integration to legacy systems, but we can't do much for those. 4. I wouldn't worry too much.

Sounds good, swapped for `CMakeCache.txt`, changed few constant names and comments.

fabiobaltieri commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,319 @@
+# opposite of the device tree inferred dependency.
+_INVERTED_PRIORITY_COMPATIBLES = set([
+        ("atmel,sam-mdio", "atmel,sam-gmac"),
+        ("ethernet-phy", "atmel,sam-gmac"),
```
Yeah was thinking about it as well, how about keeping it as a followup? Thinking let's try and work with hardcoded exception and resolve them in the devicetree first, then if we start seeing more of these come up with a way of maintaining the list. Maybe it won't be even needed at the end.

cfriedt commented:
This looks good to me, so I'll approve it when it goes green.

galak reviewed (changes requested):
Can we add some test cases for this?

galak commented:
> Can we add some test cases for this?

To clarify, I mean tests in which zephyr is built and reports the warnings of various forms.

jackrosenthal commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,316 @@
+# This is needed to load edt.pickle files.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..",
+                                "dts", "python-devicetree", "src"))
+from devicetree import edtlib  # pylint: disable=unused-import
```
@keith-zephyr and I discovered a bug where if you do this import, but use the EDT pickle, the module may not be the same between objects from the pickle and from edtlib.  e.g., you may not be able to have `except edtlib.EDTError`, as the underlying exception type objects will differ in ID.

Not sure if this affects you here but it's something to keep in mind.  Our workaround was to use `inspect.getmodule(edt_pickle)` to get the underlying module from the loaded pickle instead of importing it directly.

(Side note, pickles are usually extremely sketchy and this is why stuff like this comes out to bite!  We probably shouldn't be using them at all, but I know this pattern already existed here)

fabiobaltieri replied:
I see, I'm not doing anything with the object so I think I should be ok for this one.

jackrosenthal commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,316 @@
+
+# List of compatibles for node where the initialization priority should be the
+# opposite of the device tree inferred dependency.
+_INVERTED_PRIORITY_COMPATIBLES = set()
```
is this really a constant empty set like the `ALL_CAPS` name implies? if so, you should probably use the `frozenset()` type to make this extremely clear.

jackrosenthal commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,316 @@
+    are being initialized at which priority and stores the list internally.
+
+    Attributes:
+        file_path: path of the file to be loaded.
```
`defined_devices` isn't documented

fabiobaltieri replied:
Added a paragraph about it, lmk if it's not the correct format.

jackrosenthal commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,316 @@
+
+    def __repr__(self):
+        return (f"<ZephyrObjectFile {self.file_path} "
+                f"defined_devices: {self.defined_devices}")
```
missing `>`

fabiobaltieri replied:
Woops, sorted.

jackrosenthal commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,316 @@
+                self.defined_devices[device_ord] = prio
+
+    def __repr__(self):
+        return (f"<ZephyrObjectFile {self.file_path} "
```
`ZephyrObjectFile` ~> `self.__class__.__name__` (in case someone subclasses it)

jackrosenthal commented on scripts/build/check_init_priorities.py:
```diff
@@ -0,0 +1,316 @@
+    def __init__(self, build_dir, log):
+        self.log = log
+
+        edtser = os.path.join(build_dir, "zephyr", "edt.pickle")
```
I think this path should come from the `EDT_FILE` variable in cmake instead of hard-coding the path?

fabiobaltieri replied:
Good idea, added a flag.

galak commented:
> 

I think its worth doing something.  We need to find ways to test things that cause expected warnings or errors.

galak commented:
> I see, I could probably come up with basic simple build in `tests/` that exposes all three conditions, set it up to never run in CI (it would build fail), call it in a custom workflow step redirecting the output on a file and make sure the expected output is there. It's a bit hacky but certainly feasible, do you really want me to do it? Wondering if it's worth hassle of having an extra special test thing considering the whole script is already plenty covered in units.

I think its worth doing something. We need to find ways to test things that cause expected warnings or errors.

fabiobaltieri committed:
test: misc: add a test application for check_init_priorities

Add a test specifically for the check_init_priorities scripts.

The test creates a set of fake devices to exercise the three possible
conditions, run a build, run the check script and validate the script
output.

The check is meant to fail the build on error but that's bypassed in
this case as that would fail the twister run. That specific bit is
covered in unit tests anyway.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
> I think its worth doing something. We need to find ways to test things that cause expected warnings or errors.

Sorted!

Last update:
- cleaned up the `scripts_tests.yml` to make it a bit more readable and added `fetch-depth: 0`
- added a new patch with a full test build with fake devices with right/equal/wrong priorities and a script run to validate the exact script output

galak reviewed (commented):
I wonder about having twister supporting an expected FAIL / expected WARN.  So we don't need to do the `check_init_priorities_output`.

Wonder what others think about doing something like that.

fabiobaltieri commented:
> I wonder about having twister supporting an expected FAIL / expected WARN. So we don't need to do the `check_init_priorities_output`.
> 
> Wonder what others think about doing something like that.

Like being able to configure a test as "expected to fail the build"? Seems very specific but it would help in this case. @nashif?

Either way I'd rather avoid implementing that here, plus how it's implemented now it checks the expected output as well.

fabiobaltieri commented:
Merge conflict, rebased.

nashif commented:
> Like being able to configure a test as "expected to fail the build"? Seems very specific but it would help in this case. @nashif?
> 
> Either way I'd rather avoid implementing that here, plus how it's implemented now it checks the expected output as well.

agree, something to look into, but does not need to be part of this PR.

fabiobaltieri commented:
> This was fixed easily by deleting the entire build folder and re-running the build process so it wasn't a huge issue. I think the script should fail gracefully in this instance though rather than throwing a traceback. As mentioned, I'll work on a repro tomorrow if necessary.

That means that there are multiple object files defining the same device ordinals, the script should detect sub-builds and skip them (happens for bootloaders, sysbuild etc). If you have a sequence that reproduce this I could take a look, ideally make it work correctly.

nordicjm commented on Kconfig.zephyr:
```diff
@@ -698,6 +698,25 @@ config BUILD_OUTPUT_STRIP_PATHS
+	help
+	  Check the build for initialization priority issues by comparing the
+	  initialization priority in the build with the device dependency
+	  derived form the devicetree definition.
```
```suggestion
	  derived from the devicetree definition.
```

fabiobaltieri replied:
Fixed, thanks.

nordicjm commented on tests/misc/check_init_priorities/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: Apache-2.0
```
Typo in commit: *exercise
*but that specific bit

fabiobaltieri replied:
Sorted, thanks.

nordicjm reviewed (commented):
Very minor nits, like it!

bp-amitchone commented:
> > This was fixed easily by deleting the entire build folder and re-running the build process so it wasn't a huge issue. I think the script should fail gracefully in this instance though rather than throwing a traceback. As mentioned, I'll work on a repro tomorrow if necessary.
> 
> That means that there are multiple object files defining the same device ordinals, the script should detect sub-builds and skip them (happens for bootloaders, sysbuild etc). If you have a sequence that reproduce this I could take a look, ideally make it work correctly.

So I've got a repro for this, but it uses a custom target so let me know if you can't repro it on your end using the following steps and I'll try to put together an example project. 

I have a custom target (with custom Devicetree file, board.c etc...) based on the STM32G474; the project also supports building for the nucleo_g474re development board so that some work can be achieved using a cheap dev board rather than custom hardware. The issue occurs when these steps are followed:

1. Set the custom board as the build target
2. Reset CMake cache and re-run CMake
3. Clean and build my project (FYI using arm-none-eabi-gcc v10.3.1 20210824)
4. This build works fine, no issues
5. Then switch the target to the nucleo_g474re dev board
6. Reset CMake cache and re-run CMake
7. Clean and build project 

The error then occurs - only when building for the development board. I can't make the error occur when building for the custom target. As far as I can tell, the node that's already defined `(/soc/pin-controller@48000000/gpio@48001000)` doesn't get redefined in my custom target Devicetree file or on the nucleo_g474re's. 

So, some more information: 

CMake: v3.23.2
Python: v3.10.8
Zephyr: v3.3.0
West: v0.14.0
Dtc: v1.5.0

@martinjaeger Can you reproduce this following those steps on our dev branch?

fabiobaltieri commented:
> **Update:** Myself and @martinjaeger have attempted to reproduce this solely using west, but because west forces the use of `--pristine` when switching `-b` targets that prevents against this error. The issue must be in the way my IDE (CLion) handles cached files when it 'cleans' the build directory.
> 
> So whilst the error probably should be handled gracefully, I don't think that anyone could run into it solely using west. I even tried with a `--force` and couldn't reproduce solely using west.

Hey thanks for the details, I think that the issue is that your two boards have different device ordinals and the sequence you found results in some leftover object files that are skipped by the linking but found by the script. I did try to reproduce with west and can't for the reason you already found out. Is your IDE "clean" command leaving the `.c.obj` files there by any chance?

I can't quite detect the situation on the script and make it work properly, but I will expand the error message mentioning that there may be stale object files in the build directory and suggesting a clean.

Thanks for reporting and investigating this.

bp-amitchone commented:
> Is your IDE "clean" command leaving the `.c.obj` files there by any chance?

As far as I can tell the `.c.obj` files get deleted but CMake intermidate `.i` files and `.s` files remain. 

The 'clean' command that it runs is this: `cmake.exe --build C:\xyz\build --target clean -j 12`. I assume west's `--pristine` flag would remove these files too as it's essentially a completely fresh build directory. 

Either way, thanks for taking a look!

bp-amitchone commented:
> > As far as I can tell the `.c.obj` files get deleted but CMake intermidate `.i` files and `.s` files remain.
> > The 'clean' command that it runs is this: `cmake.exe --build C:\xyz\build --target clean -j 12`. I assume west's `--pristine` flag would remove these files too as it's essentially a completely fresh build directory.
> 
> Yeah that seems to delete the object files. Then it's really odd and I've no idea what's going on. If you feel like digging further you can try and hack the script to get the list of objects and nodes (add a `print(file, obj.defined_devices)` in line 193), that will tell you what files the script thinks are defining devices. A duplicate there would result in a duplicate symbol in the compiler so there must be two object files and only one is used for linking.
> 
> I was thinking about some introspection to this scripts if more issues like this happens but was hoping to keep it for a second stage.

Hmmm, so if I do this I do see that a couple of different devices share the same ordinals, in this case `134` and `135`: 

    gpio_stm32.c.obj {137: <Priority PRE_KERNEL_1 40>, 136: <Priority PRE_KERNEL_1 40>, 135: <Priority PRE_KERNEL_1 40>, 134: <Priority PRE_KERNEL_1 40>, 26: <Priority PRE_KERNEL_1 40>, 87: <Priority PRE_KERNEL_1 40>, 29: <Priority PRE_KERNEL_1 40>}

    eeprom_at2x.c.obj {135: <Priority POST_KERNEL 80>, 134: <Priority POST_KERNEL 80>}

It's easy enough to detect duplicates (albeit crudely):

    if obj.defined_devices:
        for k, v in obj.defined_devices.items():
            if k not in ords_list:
                ords_list.append(k)
            else:
                dups_list.append([k, v])

So I wonder if it's worth doing so and then exiting with an error code and reporting back to the user that they probably need to do a `--pristine` or delete their build directory and rebuild?

fabiobaltieri commented:
> Hmmm, so if I do this I do see that a couple of different devices share the same ordinals, in this case `134` and `135`:
> 
> ```
> gpio_stm32.c.obj {137: <Priority PRE_KERNEL_1 40>, 136: <Priority PRE_KERNEL_1 40>, 135: <Priority PRE_KERNEL_1 40>, 134: <Priority PRE_KERNEL_1 40>, 26: <Priority PRE_KERNEL_1 40>, 87: <Priority PRE_KERNEL_1 40>, 29: <Priority PRE_KERNEL_1 40>}
> 
> eeprom_at2x.c.obj {135: <Priority POST_KERNEL 80>, 134: <Priority POST_KERNEL 80>}
> ```

Can you check what 134 and 135 are in the two builds? (check in ./zephyr/include/generated/devicetree_generated.h) Is it possible that `gpio_at2x` is only part of the custom board and is not linked in the nucleo build? In that case it should not be there after the clean-rebuild.

fabiobaltieri commented:
Hi, back working on it, updates in this push

- rebased
- fixed few typos
- improved the "already defined" error message, suggesting to clean the build dir
- rewrote the twister test check to not use any unix utility, just added a small python script in the test directory to do the check, the reference output is embedded in there

henrikbrixandersen commented on Kconfig.zephyr:
```diff
@@ -717,6 +717,25 @@ config BUILD_OUTPUT_STRIP_PATHS
 	  Debuggers usually have a path mapping feature to ensure the files are
 	  still found.
 
+config CHECK_INIT_PRIORITIES
+	bool "Build time initialization priorities check"
+	help
+	  Check the build for initialization priority issues by comparing the
+	  initialization priority in the build with the device dependency
+	  derived from the devicetree definition.
```
At some point, we need to make this `default y`.

fabiobaltieri replied:
Yeah, next step for this is to do a  `default y` CI run and start engaging with the maintainers of whatever fails.

nordicjm reviewed (dismissed):
Great! Be sure to submit a release note update PR once this is merged

tejlmand reviewed (changes requested):
looks good, but I notice an additional commit with test application (which is very good 👍 ).

But the invocation of the verification script need a small improvement.

fabiobaltieri commented on tests/misc/check_init_priorities/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+
+add_custom_target(
+	check_init_priorities_output
+	COMMAND echo "Running check_init_priorities.py"
```
Done.

fabiobaltieri commented on tests/misc/check_init_priorities/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+	COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/build/check_init_priorities.py
+	  --verbose
+	  --build-dir ${PROJECT_BINARY_DIR}/.. 2>&1
+	  | ${APPLICATION_SOURCE_DIR}/validate_check_init_priorities_output.py
```
Ok cool, I added a flag to the core script to add logging into a file and changed the test script to read a file instead.

fabiobaltieri commented:
Few more updates:
- added two flags to the core script, a `--output` to redirect the output to a file and a `--always-succeed` to always return with a 0 error code, these to support the testcase without potentially unportable CMake situations
- changed the test script to get the script output from a file instead of stdin

henrikbrixandersen reviewed (approved):
LGTM, thanks.

carlescufi reviewed (changes requested):
Before I merge this I want to get @stephanosio and @mbolivar-nordic's opinion regarding the potential merging of the new workflow with this one:
https://github.com/zephyrproject-rtos/zephyr/blob/e51a90934344dbae214cdd94a6130ce35246a550/.github/workflows/west_cmds.yml#L80

gmarull commented on .github/workflows/scripts_tests.yml:
```diff
@@ -0,0 +1,53 @@
+# Copyright 2023 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+name: Scripts tests
```
I'm all in to standardize a bit testing of scripts/, I actually failed a long time ago to do something similar https://github.com/zephyrproject-rtos/zephyr/pull/34435. Next would be formatting, which is now a mess.

fabiobaltieri commented:
> Before I merge this I want to get @stephanosio and @mbolivar-nordic's opinion regarding the potential merging of the new workflow with this one:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/e51a90934344dbae214cdd94a6130ce35246a550/.github/workflows/west_cmds.yml#L80

Could we handle it as a followup? Would love to get this in before feature freeze, I could also pull the workflow patch and we can handle CI integration separately if the conversation is just about it.
